### ISSUE TITLE
Encoding Error / Mix "Mysql2::Error: Illegal mix of collations"

### ISSUE THREAD

**Issue by Andreas3C** [danmunn/redmine_dmsf#1420]
Good day,

thank you very much for the help in #1406 , with Redmine 5.0 on a new server I could get further with your help. :)

Unfortunately I have a new problem. When I click on the tab "DMS" in Redmine, I get an Internal Server Error 500.
In the log I could find a problem with the encoding, a mix of UTF8 and Latin1. When researching I was able to find the following:
#820 

There it is assumed that the Redmine DB was set up incorrectly (not as UTF8). However, I had it installed as UTF8 in my case. Also a new try with UTF8MB4 did not bring any change.

Now I checked our old server, from which I want to transfer all data, and it also has a mix of encodings in it - but there everything works... (in contrast to my new server with Redmine 5.0).
I am now unsure what to do. My DB is ok, the new DB was also ok. But somewhere in the DMSF area it says Latin, and that has to go.... are there any ideas for this?

Thanks!

LG
Andreas

Old Server (running):
```
mysql> SELECT @@character_set_database, @@collation_database;
+--------------------------+----------------------+
| @@character_set_database | @@collation_database |
+--------------------------+----------------------+
| utf8                     | utf8_general_ci      |
+--------------------------+----------------------+
1 row in set (0,00 sec)

mysql> show variables like "character_set_database";
+------------------------+-------+
| Variable_name          | Value |
+------------------------+-------+
| character_set_database | utf8  |
+------------------------+-------+
1 row in set (0,01 sec)

mysql> show create table dmsf_file_revisions;
+---------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table               | Create Table                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
+---------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| dmsf_file_revisions | CREATE TABLE `dmsf_file_revisions` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `dmsf_file_id` int(11) NOT NULL,
  `source_dmsf_file_revision_id` int(11) DEFAULT NULL,
  `name` varchar(255) NOT NULL,
  `disk_filename` varchar(255) NOT NULL,
  `size` bigint(20) DEFAULT NULL,
  `mime_type` varchar(255) DEFAULT NULL,
  `title` varchar(255) NOT NULL,
  `description` text,
  `workflow` int(11) DEFAULT NULL,
  `major_version` int(11) NOT NULL,
  `minor_version` int(11) NOT NULL,
  `comment` text,
  `deleted` tinyint(1) NOT NULL DEFAULT '0',
  `deleted_by_user_id` int(11) DEFAULT NULL,
  `user_id` int(11) NOT NULL,
  `created_at` datetime NOT NULL,
  `updated_at` datetime NOT NULL,
  `dmsf_workflow_id` int(11) DEFAULT NULL,
  `dmsf_workflow_assigned_by` int(11) DEFAULT NULL,
  `dmsf_workflow_assigned_at` datetime DEFAULT NULL,
  `dmsf_workflow_started_by` int(11) DEFAULT NULL,
  `dmsf_workflow_started_at` datetime DEFAULT NULL,
  PRIMARY KEY (`id`),
  KEY `index_dmsf_file_revisions_on_dmsf_file_id` (`dmsf_file_id`)
) ENGINE=InnoDB AUTO_INCREMENT=5892 DEFAULT CHARSET=latin1 |
+---------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0,00 sec)

mysql>
```

**Comment by Andreas3C** [comment]
Thank you for your answer!

I am unsure how to enter the block you mentioned. Line by line? Or all in one line? Or completely different? Sorry I don't understand, until now I always had single line commands with a ; at the end.

What is my TABLE_NAME, is it "dmsf_file_revisions" or are there other tables from DMSF that need to be adjusted? Could I also do it via the following command (if my TABLE_NAME is correct):

`ALTER TABLE dmsf_file_revisions CONVERT TO CHARACTER SET utf8mb4;`

**Comment by picman** [comment]
If you run the recommended command, e.g.
```sql
SELECT
    CONCAT(
        'ALTER TABLE `',
        TABLE_NAME,
        '` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci;'
    ) AS mySQL
FROM
    INFORMATION_SCHEMA. TABLES
WHERE
    TABLE_SCHEMA = "redmine"
AND TABLE_TYPE = "BASE TABLE";
```
you will get several SQL commands as the output. Something like this:
```sql
ALTER TABLE `journal_details` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci;
ALTER TABLE `imports` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci; 
...
78 rows in set (0,002 sec)
```
Then you can either run these commands one by one or, and that I find more practical, run all of them at once.
You can also put them into a text file and then to to process this file, e.g.:
```bash
sudo mysql redmine < your_file.sql
```

**Comment by Andreas3C** [comment]
Thank you!
But I'm still unsure, which parameters have to be adjusted in the command. If I have to adjust "TABLE_NAME", wich name do I have to use? Is it dmsf_file_revisions?
My DB has the name redmine5. Do I have to adjust the "TABLE_SCHEMA" from redmine to redmine5?

Thank you!

**Comment by picman** [comment]
Adjust just `TABLE_SCHEMA="redmine5"`.

**Comment by picman** [comment]
That's strange. On my computer, I have:
```sql
show create table dmsf_file_revisions;
...
ENGINE=InnoDB AUTO_INCREMENT=322753 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
```
So, collation `utf8mb4_unicode_ci`.
If I change it to `general`:
```sql
ALTER TABLE `dmsf_file_revisions` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci;
```
I see then `unicode`:
```sql
show create table dmsf_file_revisions;
...
ENGINE=InnoDB AUTO_INCREMENT=322753 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_general_ci
```
It changes the collation from `unicode` to `general` as requested.

Maybe, try also to change the collation for the database.

```sql
ALTER DATABASE redmine5 CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci;
ALTER TABLE `dmsf_file_revisions` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci;
```

**Comment by Andreas3C** [comment]
Strange, when I do it manually, it works:

```
mysql> ALTER TABLE `dmsf_file_revisions` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci;
Query OK, 5866 rows affected (0.34 sec)
Records: 5866  Duplicates: 0  Warnings: 0

mysql> show create table dmsf_file_revisions;
+---------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table               | Create Table                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
+---------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| dmsf_file_revisions | CREATE TABLE `dmsf_file_revisions` (
  `id` int NOT NULL AUTO_INCREMENT,
  `dmsf_file_id` int NOT NULL,
  `source_dmsf_file_revision_id` int DEFAULT NULL,
  `name` varchar(255) COLLATE utf8mb4_general_ci NOT NULL,
  `disk_filename` varchar(255) COLLATE utf8mb4_general_ci NOT NULL,
  `size` bigint DEFAULT NULL,
  `mime_type` varchar(255) COLLATE utf8mb4_general_ci DEFAULT NULL,
  `title` varchar(255) COLLATE utf8mb4_general_ci NOT NULL,
  `description` mediumtext COLLATE utf8mb4_general_ci,
  `workflow` int DEFAULT NULL,
  `major_version` int NOT NULL,
  `minor_version` int NOT NULL,
  `comment` mediumtext COLLATE utf8mb4_general_ci,
  `deleted` tinyint(1) NOT NULL DEFAULT '0',
  `deleted_by_user_id` int DEFAULT NULL,
  `user_id` int NOT NULL,
  `created_at` datetime NOT NULL,
  `updated_at` datetime NOT NULL,
  `dmsf_workflow_id` int DEFAULT NULL,
  `dmsf_workflow_assigned_by` int DEFAULT NULL,
  `dmsf_workflow_assigned_at` datetime DEFAULT NULL,
  `dmsf_workflow_started_by` int DEFAULT NULL,
  `dmsf_workflow_started_at` datetime DEFAULT NULL,
  PRIMARY KEY (`id`),
  KEY `index_dmsf_file_revisions_on_dmsf_file_id` (`dmsf_file_id`)
) ENGINE=InnoDB AUTO_INCREMENT=5892 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_general_ci |
+---------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql>
```

So I'm wondering, if I should do it for all tables manually... ;)

**Comment by picman** [comment]
You can change your SQL script for all tables into a shell script which will consists of the following commands:
```shell
mysql redmine5 -e "ALTER TABLE `dmsf_file_revisions` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci;"
...
```

**Comment by Andreas3C** [comment]
I did not get the shell script to work, but that may be due to my lack of experience ;).
I have now simply written all commands in a row and executed in msql. 

I had already executed the fix from #1406 . Then I installed the plugin according to the instructions and it now works fine. Thanks a lot for your help!

**Comment by picman** [comment]
Great!
